File: ios/ILoveTrains/Core/Recommendation.swift
import Foundation

let transferPenaltyMillis: Millis = 300_000

private struct RecommendationTuple {
    var cost: Millis
    var changes: Int
    var arrival: Millis
    var departure: Millis
    var identity: [(String, Millis)]
}

func recommendationCost(_ journey: Journey) -> Millis? {
    recommendationTuple(journey)?.cost
}

func compareRecommendations(_ left: Journey, _ right: Journey) -> ComparisonResult {
    guard let lhs = recommendationTuple(left), let rhs = recommendationTuple(right) else {
        return recommendationTuple(left) != nil ? .orderedAscending
            : recommendationTuple(right) != nil ? .orderedDescending : .orderedSame
    }
    for (a, b) in [(lhs.cost, rhs.cost), (Millis(lhs.changes), Millis(rhs.changes)),
                   (lhs.arrival, rhs.arrival), (lhs.departure, rhs.departure)] {
        if a != b { return a < b ? .orderedAscending : .orderedDescending }
    }
    for (a, b) in zip(lhs.identity, rhs.identity) {
        let name = compareCodePoints(a.0, b.0)
        if name != .orderedSame { return name }
        if a.1 != b.1 { return a.1 < b.1 ? .orderedAscending : .orderedDescending }
    }
    if lhs.identity.count == rhs.identity.count { return .orderedSame }
    return lhs.identity.count < rhs.identity.count ? .orderedAscending : .orderedDescending
}

func selectRecommendation(
    _ journeys: [Journey],
    now: Millis,
    modes: Set<String> = allModes,
    maxTransfers: Int? = nil
) -> Journey? {
    journeys
        .filter { recommendationEligible($0, now: now, modes: modes, maxTransfers: maxTransfers) }
        .min { compareRecommendations($0, $1) == .orderedAscending }
}

func recommendationIsFresh(_ value: JourneyRecommendation, now: Millis, maxTransfers: Int?) -> Bool {
    value.board.requestMaxTransfers == maxTransfers && value.board.isLive(now) && value.journey.retained != true
}

func selectRecommendation(
    _ candidates: [JourneyRecommendation],
    now: Millis,
    modes: Set<String>,
    maxTransfers: Int?
) -> JourneyRecommendation? {
    let allowed = candidates.filter {
        recommendationEligible($0.journey, now: now, modes: modes, maxTransfers: maxTransfers)
    }
    let fresh = allowed.filter {
        $0.board.requestMaxTransfers == maxTransfers && $0.board.isLive(now) && $0.journey.retained != true
    }
    return (fresh.isEmpty ? allowed : fresh).min {
        compareRecommendations($0.journey, $1.journey) == .orderedAscending
    }
}

func recommendationCandidates(_ board: BoardData) -> [JourneyRecommendation] {
    var values: [String: JourneyRecommendation] = [:]
    func add(_ source: BoardData) {
        for journey in source.journeys {
            let candidate = JourneyRecommendation(journey: journey, board: source)
            if let existing = values[journey.key], existing.board.generatedAt >= source.generatedAt { continue }
            values[journey.key] = candidate
        }
    }
    add(board)
    for page in board.recommendationPages ?? [] { add(page.board(from: board.from, to: board.to)) }
    if let recommendation = board.recommendation { add(recommendation.board(from: board.from, to: board.to)) }
    return Array(values.values)
}

func earliestAlternative(
    _ candidates: [JourneyRecommendation],
    recommended: Journey,
    now: Millis,
    modes: Set<String>,
    maxTransfers: Int?
) -> JourneyRecommendation? {
    guard let firstIdentity = recommended.legs.first.map({ ($0.line, $0.departure) }) else { return nil }
    return candidates.filter {
        guard recommendationEligible($0.journey, now: now, modes: modes, maxTransfers: maxTransfers),
              let leg = $0.journey.legs.first else { return false }
        return leg.line != firstIdentity.0 || leg.departure != firstIdentity.1
    }.min {
        if $0.journey.effectiveDeparture != $1.journey.effectiveDeparture {
            return $0.journey.effectiveDeparture < $1.journey.effectiveDeparture
        }
        return compareRecommendations($0.journey, $1.journey) == .orderedAscending
    }
}

func nextRecommendationCursor(
    page: BoardData,
    previousAt: Millis,
    now: Millis,
    best: Journey?
) -> Millis? {
    guard let greatest = page.journeys.map(\.effectiveDeparture).filter(\.isFinite).max() else { return nil }
    let cursor = (floor(greatest / 600_000) + 1) * 600_000
    if cursor <= previousAt || cursor > now + 7_200_000 { return nil }
    if let bestCost = best.flatMap(recommendationCost), cursor > bestCost { return nil }
    return cursor
}

private func recommendationEligible(
    _ journey: Journey,
    now: Millis,
    modes: Set<String>,
    maxTransfers: Int?
) -> Bool {
    guard recommendationTuple(journey) != nil,
          journey.effectiveDeparture >= now,
          journeyAllowed(journey, modes: modes) else { return false }
    return maxTransfers.map { journey.legs.count - 1 <= $0 } ?? true
}

private func recommendationTuple(_ journey: Journey) -> RecommendationTuple? {
    guard !journey.legs.isEmpty, !journey.cancelled else { return nil }
    for leg in journey.legs {
        guard validTransitMillis(leg.departure), validTransitMillis(leg.arrival),
              validTransitMillis(leg.effectiveDeparture), validTransitMillis(leg.effectiveArrival),
              leg.effectiveArrival >= leg.effectiveDeparture else { return nil }
    }
    for index in 1..<journey.legs.count where
        journey.legs[index].effectiveDeparture < journey.legs[index - 1].effectiveArrival {
        return nil
    }
    let departure = journey.effectiveDeparture
    let arrival = journey.effectiveArrival
    guard arrival >= departure else { return nil }
    let changes = journey.legs.count - 1
    return RecommendationTuple(
        cost: arrival + Millis(changes) * transferPenaltyMillis,
        changes: changes,
        arrival: arrival,
        departure: departure,
        identity: journey.legs.map { ($0.line, $0.departure) }
    )
}

private func compareCodePoints(_ left: String, _ right: String) -> ComparisonResult {
    let lhs = left.unicodeScalars.map(\.value)
    let rhs = right.unicodeScalars.map(\.value)
    for (a, b) in zip(lhs, rhs) where a != b {
        return a < b ? .orderedAscending : .orderedDescending
    }
    if lhs.count == rhs.count { return .orderedSame }
    return lhs.count < rhs.count ? .orderedAscending : .orderedDescending
}

File: android/app/src/main/java/com/ilovetrains/app/Recommendation.kt
package com.ilovetrains.app

const val TransferPenaltyMillis = 300_000L
private const val RecommendationBucketMillis = 600_000L
private const val RecommendationFutureLimitMillis = 7_200_000L

data class RecommendationCandidate(val journey: Journey, val source: BoardData, val stale: Boolean)

fun recommendationCost(journey: Journey): Long? = recommendationTuple(journey)?.cost

fun compareRecommendations(first: Journey, second: Journey): Int {
    val left = recommendationTuple(first)
    val right = recommendationTuple(second)
    if (left == null || right == null) return if (left != null) -1 else if (right != null) 1 else 0
    compareValues(left.cost, right.cost).takeIf { it != 0 }?.let { return it }
    compareValues(left.changes, right.changes).takeIf { it != 0 }?.let { return it }
    compareValues(left.arrival, right.arrival).takeIf { it != 0 }?.let { return it }
    compareValues(left.departure, right.departure).takeIf { it != 0 }?.let { return it }
    for (index in 0 until minOf(left.identity.size, right.identity.size)) {
        compareCodePoints(left.identity[index].first, right.identity[index].first).takeIf { it != 0 }?.let { return it }
        compareValues(left.identity[index].second, right.identity[index].second).takeIf { it != 0 }?.let { return it }
    }
    return compareValues(left.identity.size, right.identity.size)
}

fun selectRecommendation(
    candidates: List<RecommendationCandidate>,
    now: Long,
    modes: Set<String> = AllModes,
    maxTransfers: Int? = null,
): RecommendationCandidate? {
    val eligible = candidates.filter { candidate ->
        recommendationTuple(candidate.journey) != null && candidate.journey.effectiveDeparture >= now &&
            journeyAllowed(candidate.journey, modes) && candidate.journey.withinTransferCap(maxTransfers)
    }
    val pool = eligible.filterNot(RecommendationCandidate::stale).ifEmpty { eligible }
    return pool.minWithOrNull { a, b -> compareRecommendations(a.journey, b.journey) }
}

fun earliestAlternative(
    candidates: List<RecommendationCandidate>,
    recommended: Journey,
    now: Long,
    modes: Set<String> = AllModes,
    maxTransfers: Int? = null,
): RecommendationCandidate? {
    val identity = firstServiceIdentity(recommended)
    return candidates.asSequence().filter { candidate ->
        val journey = candidate.journey
        recommendationTuple(journey) != null && journey.effectiveDeparture >= now &&
            journeyAllowed(journey, modes) && journey.withinTransferCap(maxTransfers) && firstServiceIdentity(journey) != identity
    }.minWithOrNull { a, b ->
        compareValues(a.journey.effectiveDeparture, b.journey.effectiveDeparture).takeIf { it != 0 }
            ?: compareRecommendations(a.journey, b.journey)
    }
}

fun nextRecommendationCursor(page: BoardData, previousAt: Long, now: Long, best: Journey?): Long? {
    val greatest = page.journeys.maxOfOrNull(Journey::effectiveDeparture) ?: return null
    val cursor = (greatest / RecommendationBucketMillis + 1) * RecommendationBucketMillis
    val cost = best?.let(::recommendationCost)
    return cursor.takeIf { it > previousAt && it <= now + RecommendationFutureLimitMillis && (cost == null || it <= cost) }
}

fun BoardData.recommendationCandidates(now: Long, current: TransferConstraint? = fetchConstraint): List<RecommendationCandidate> {
    val rows = linkedMapOf<String, RecommendationCandidate>()
    fun add(source: BoardData, stale: Boolean) {
        source.journeys.forEach { journey ->
            val candidate = RecommendationCandidate(journey, source, stale || !source.isLive(now) || journey.retained)
            val previous = rows[journey.key]
            if (previous == null || source.generatedAt > previous.source.generatedAt ||
                source.generatedAt == previous.source.generatedAt && previous.stale && !candidate.stale) rows[journey.key] = candidate
        }
    }
    add(this, fetchConstraint != current || offline || serverStale)
    recommendationPages.forEach { page ->
        add(page.body, page.constraint != current || page.serverStale || page.body.offline)
    }
    recommendation?.let { result ->
        add(result.source.copy(journeys = listOf(result.journey)), result.source.fetchConstraint != current)
    }
    return rows.values.toList()
}

private data class RecommendationTuple(
    val cost: Long,
    val changes: Int,
    val arrival: Long,
    val departure: Long,
    val identity: List<Pair<String, Long>>,
)

private fun recommendationTuple(journey: Journey): RecommendationTuple? {
    if (journey.legs.isEmpty() || journey.cancelled) return null
    if (journey.legs.any { it.cancelled || it.arrival < it.departure || it.effectiveArrival < it.effectiveDeparture }) return null
    if (journey.legs.zipWithNext().any { (previous, next) -> next.effectiveDeparture < previous.effectiveArrival }) return null
    val arrival = journey.effectiveArrival
    val departure = journey.effectiveDeparture
    if (arrival < departure) return null
    val changes = journey.legs.size - 1
    val penalty = runCatching { Math.multiplyExact(changes.toLong(), TransferPenaltyMillis) }.getOrNull() ?: return null
    val cost = runCatching { Math.addExact(arrival, penalty) }.getOrNull() ?: return null
    return RecommendationTuple(cost, changes, arrival, departure, journey.legs.map { it.line to it.departure })
}

private fun firstServiceIdentity(journey: Journey): Pair<String, Long>? = journey.legs.firstOrNull()?.let { it.line to it.departure }

internal fun compareCodePoints(first: String, second: String): Int {
    var a = 0
    var b = 0
    while (a < first.length && b < second.length) {
        val left = first.codePointAt(a)
        val right = second.codePointAt(b)
        if (left != right) return compareValues(left, right)
        a += Character.charCount(left)
        b += Character.charCount(right)
    }
    return compareValues(first.codePointCount(0, first.length), second.codePointCount(0, second.length))
}
